---
title: "Multiple Peaks"
author: "Simone Zaghen"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document: 
    toc: true
    number_section: true
    toc_float: 
      collapsed: false
      smooth_scroll: true
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "../results/") })

---

After running the snakemake pipeline, this file can be used to extract interesting genes and peaks from the output data. 
In this script, I first import data and metadata (the script is set up for working on one file at the time right now, but can easily be scaled to all files using the map/map2 functions from the purrr package) and merge them. I then try to extract the most interesting genes/peaks and the visualize them on IGV.

# Load packages

```{r, message=FALSE}
library(tidyverse)

```

# Load data and merge metadata

```{r}
rawdata <- read.delim("/scratch/sz4633/polyadenylation_cerevisiae_split_strands/peaks_area/RAPA1_negative_PeakArea.tsv")

metadata <- read.delim("/scratch/sz4633/polyadenylation_cerevisiae_split_strands/peaks_filtered/RAPA1_negative_full.tsv")

df <- merge(rawdata, metadata, by = c("chromosome",
                                     "peak_start",
                                     "peak_end",
                                     "peak_name",
                                     "gene_name"))

```

Check how many peaks are assigned to genes. Remember that in previous steps of the pipeline I removed genes with only one peak assigned. I did that because it speeds up calculating the coverage. 

```{r}
tmp <- as.data.frame(table(df$gene_name))
tmp <- as.data.frame(table(tmp$Freq))
colnames(tmp) <- c("# of peaks", "Freq")
tmp  

```

Calculate the sd of the peaks area assigned to each of the genes. The sd expresses which genes have most similar/different peak areas. I also calculate how many peaks are assigned to each gene. 

```{r}
df <- df %>% 
  group_by(gene_name) %>%
  mutate(st_dev = sd(peak_area), #calculate difference between peak areas
         freq = n()) %>% #calculate how many peaks are assigned to each gene
  ungroup()

```

Then extract top 100 genes for area, and sort them by their sd and distance from stop codon. Then I can check some of these genes on IGV.

```{r}
#extract top 100 genes for area
interesting_genes <- unique(df %>%
                              arrange(desc(peak_area)) %>%
                              pull(gene_name))[1:50]
#sort them
interesting_genes <- df %>%
  filter(gene_name %in% interesting_genes) %>%
  arrange(st_dev, peak_to_stop_codon, gene_name)

```

